### ISSUE TITLE
Replace `sed` in favor of a simpler `replace` string that works like `clojure.string/replace`

### ISSUE THREAD

**Issue by devth** [yetibot/yetibot#702]
Sed has problems with strings that contain `/` and I'd rather not support necessary escaping.

**Comment by devth** [comment]
hey @PabloReszczynski! 

1. Checkout the [existing sed implementation](https://github.com/yetibot/yetibot/blob/master/src/yetibot/commands/sed.clj). It's pretty simple and uses `clojure.string/replace` already.
2. Come up with a new command (maybe we call it `replace`?) that basically exposes `replace`'s functionality. API isn't defined so it's up to you/us to figure out what it should look like. I'd say we retain passing the actual string to replace via piped contents (i.e. keep using `raw` as the string to replace). That means we just need to parse out 2 params. The existing regex `#"^s\/(.+)\/(.*)"` expects them to be separated by `/` but we should probably drop that syntax. Could use whitespace separated (meaning the pattern to match can't have any whitespace but people can use `\s` or `.` instead).

`clojure.string/replace` docstring for reference:

```
clojure.string/replace
[s match replacement]
  Replaces all instance of match with replacement in s.

   match/replacement can be:

   string / string
   char / char
   pattern / (string or function of match).

   See also replace-first.

   The replacement is literal (i.e. none of its characters are treated
   specially) for all cases above except pattern / string.

   For pattern / string, $1, $2, etc. in the replacement string are
   substituted with the string that matched the corresponding
   parenthesized group in the pattern.  If you wish your replacement
   string r to be used literally, use (re-quote-replacement r) as the
   replacement argument.  See also documentation for
   java.util.regex.Matcher's appendReplacement method.

   Example:
   (clojure.string/replace "Almost Pig Latin" #"\b(\w)(\w+)\b" "$2$1ay")
   -> "lmostAay igPay atinLay"
```

Let me know if you have questions! Post here or join #dev room in [Slack](https://slack.yetibot.com) if you want to chat.

**Comment by PabloReszczynski** [comment]
Having made a simple replace command that only work with strings, I am unsure how to proceed, syntax wise, to allow regex patterns and functions (as the `clojure.string/replace` API allows).

**Comment by devth** [comment]
What do you think about this?

```clojure
(cmd-hook #"replace"
          #"(\S+)\s+(\S)" replace-cmd)
```

Captures two non-whitespace groups of characters separated by whitespace.

**Comment by PabloReszczynski** [comment]
Is there a way I can unit test the high level command API? Just to see if running "replace foobar foo bar" works as expected and without using directly `replace-cmd`.

**Comment by devth** [comment]
Not currently but I've often thought it'd be useful. Let me think on that a bit and maybe I can add something for that today or soon. 🤔

**Comment by devth** [comment]
I often manually test commands in Slack as the last step but agree some test automation would be a little easier.

**Comment by PabloReszczynski** [comment]
That would be very nice!
Either way I'll be committing the first version of the command soon on my fork so you can take a look at it.

**Comment by devth** [comment]
@PabloReszczynski I'm tracking the issue to improve command testing in https://github.com/yetibot/yetibot/issues/745 FYI! I have a bit of untangling / decomplecting to do to get it working so it might take more than a day or two.

**Comment by devth** [comment]
Hey @PabloReszczynski check out a sample of the data returned so far: https://github.com/yetibot/yetibot/issues/745#issuecomment-429120367

How does that look for improving testability of commands?

**Comment by PabloReszczynski** [comment]
Looks great!
I have a question regarding the intended API of `replace`. As for now it works as a simple search and replace function, but `clojure.string/replace` also allows to match regex and apply functions.
For regex, I think using the same clojure pattern for regex would work.
But I don't know about functions. Should we allow functions in commands at all? Or maybe have some options like `-U` for uppercase, `-L` for lowercase, etc

**Comment by devth** [comment]
Yeah, it seems awkward to embed a Clojure function in a Yetibot command, so I would skip that. But regex would be fine I think.

**Comment by devth** [comment]
Hey @PabloReszczynski, any other questions? I needed this feature the other day so came back to see if you were blocked on anything 😁
